src/worldcereal/utils/upload.py

```diff
@@ -0,0 +1,213 @@
+        Takes an OpenEO connection object and returns temporary credentials to interact with S3
+        """
+        auth_token = conn.auth.bearer.split("/")
+        os.environ["AWS_ENDPOINT_URL_STS"] = cls.STS_ENDPOINT
+        sts = boto3.client("sts")
+        return AWSSTSCredentials._from_assume_role_response(
+            sts.assume_role_with_web_identity(
+                RoleArn="arn:aws:iam::000000000000:role/S3Access",
+                RoleSessionName=auth_token[1],
+                WebIdentityToken=auth_token[2],
+                DurationSeconds=43200,
+            )
+        )
+
+    def get_user_hash(self) -> str:
+        hash_object = hashlib.sha1(self.subject_from_web_identity_token.encode())
+        return hash_object.hexdigest()
+
+
+class OpenEOArtifactHelper:
+    BUCKET_NAME = "OpenEO-artifacts"
+    S3_ENDPOINT = "https://s3.prod.warsaw.openeo.dataspace.copernicus.eu"
+    # From what size will we switch to multi-part-upload
+    MULTIPART_THRESHOLD_IN_MB = 50
+
+    def __init__(self, creds: AWSSTSCredentials):
+        self._creds = creds
+        self.session = boto3.Session(**creds.as_kwargs())
+
+    @classmethod
+    def from_openeo_connection(cls, conn: Connection) -> OpenEOArtifactHelper:
+        creds = AWSSTSCredentials.from_openeo_connection(conn)
+        return OpenEOArtifactHelper(creds)
+
+    def get_s3_client(self):
+        return self.session.client("s3", endpoint_url=self.S3_ENDPOINT)
+
+    def set_env(self):
+        os.environ["AWS_ENDPOINT_URL_S3"] = self.S3_ENDPOINT
+
+    def user_prefix(self) -> str:
+        """Each user has its own prefix retrieve it"""
+        return self._creds.get_user_hash()
+
+    def get_upload_prefix(self):
+        return (
+            f"{self.user_prefix()}/{datetime.datetime.utcnow().strftime('%Y/%m/%d')}/"
+        )
+
+    def get_upload_key(self, object_name: str) -> str:
+        return f"{self.get_upload_prefix()}{object_name}"
+
+    def upload_bytes(self, object_name: str, blob: bytes) -> str:
+        """Upload a bunch of bytes into an object and return an S3 URI to it"""
+        bucket = self.BUCKET_NAME
+        key = self.get_upload_key(object_name)
+        self.get_s3_client().put_object(Body=blob, Bucket=bucket, Key=key)
+        return f"s3://{bucket}/{key}"
+
+    def upload_string(self, object_name: str, s: str, encoding: str = "utf-8") -> str:
+        """Upload a string into an object"""
+        return self.upload_bytes(object_name, s.encode(encoding))
+
+    def upload_file(self, object_name: str, src_file_path: str) -> str:
+        MB = 1024**2
+        config = TransferConfig(multipart_threshold=self.MULTIPART_THRESHOLD_IN_MB * MB)
+
+        bucket = self.BUCKET_NAME
+        key = self.get_upload_key(object_name)
+
+        self.get_s3_client().upload_file(src_file_path, bucket, key, Config=config)
+        return f"s3://{bucket}/{key}"
+
+    def get_presigned_url(
+        self, s3_uri: str, expires_in_seconds: int = 3600 * 24
```

Comment:
You could increase the lifetime of the presigned URL to 6 days rather than 1 day.

Reply:
Done in [c0b2140](https://github.com/WorldCereal/worldcereal-classification/pull/170/commits/c0b2140be4aedf526b8b9cc70af78b2cfc4ee132)